smolyay/basis.py

```diff
@@ -0,0 +1,226 @@
+import abc
+import math
+
+import numpy
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @abc.abstractproperty
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Compute term of basis function
+        Returns the output of the basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis function Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a sequence of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    ``n`` is the degree of the polynomial this object describes
+    ``points`` represents the extrema of the polynomial of this degree on
+    the domain [-1,1]
+
+    The extrema of the polynomial are calculated via the following
+    equation:
+
+    ..math:
+        x_{n,j}^* = -\cos\left((\frac{j-1}{n-1}\pi\right), j = 1,...,n
+        n = 2^{exactness-1}+1
+
+    Parameters
+    ----------
+    n : int
+        degree of the Chebyshev polynomial
+    """
+
+    def __init__(self,n):
+        super().__init__()
+        self._n = n
+
+    @property
+    def points(self):
+        """extrema of polynomial"""
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    @n.setter
+    def n(self,n):
+        if self._n != n:
+            self._n = n
+            self._points = []
+
+    def _compute_points(self):
+        """Compute extrema of Chebyshev polynomial of the first kind"""
+        if self._n == 0:
+            self._points = [0]
+        else:
+            i = numpy.arange(self._n+1)
+            self._points = list(-numpy.cos(numpy.pi*i/self._n))
+
+    def __call__(self,x):
```

Comment:
Just making a note that we may want to think about how this calculation is done later. Mohammadreza found a method using branching, and I read that NumPy evaluates Chebyshev polynomials using something called Clenshaw recursion:
https://numpy.org/doc/stable/reference/generated/numpy.polynomial.chebyshev.chebval.html#numpy.polynomial.chebyshev.chebval
https://en.wikipedia.org/wiki/Clenshaw_algorithm
I don't know if the later helps here or not, since we have only one term in the series.